tracker/watcher.go

```diff
@@ -6,74 +6,96 @@ SPDX-License-Identifier: Apache-2.0
 package tracker
 
 import (
-	"fmt"
+	"context"
 	"sync"
 	"time"
 
 	"github.com/go-logr/logr"
 	"github.com/vmware-labs/reconciler-runtime/client"
-	"github.com/vmware-labs/reconciler-runtime/inject"
+	"github.com/vmware-labs/reconciler-runtime/manager"
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/runtime/schema"
 	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
 	crclient "sigs.k8s.io/controller-runtime/pkg/client"
-	"sigs.k8s.io/controller-runtime/pkg/controller"
 	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 	"sigs.k8s.io/controller-runtime/pkg/source"
 )
 
-type watchFunc func(gvk schema.GroupVersionKind, controller controller.Controller) error
+type watchFunc func(gvk schema.GroupVersionKind) (context.CancelFunc, error)
 
 type watcher struct {
-	tracker Tracker
+	tracker GKAwareTracker
 	watch   watchFunc
 
-	m          sync.Mutex // protects watches and controller
-	watches    map[schema.GroupKind]struct{}
-	controller controller.Controller
+	m       sync.Mutex // protects watches
+	watches map[schema.GroupKind]context.CancelFunc
 }
 
 var (
-	_ Tracker           = (*watcher)(nil)
-	_ inject.Controller = (*watcher)(nil)
+	_ Tracker = (*watcher)(nil)
 )
 
+var nopReconciler = reconcile.Func(func(context.Context, reconcile.Request) (reconcile.Result, error) {
+	return reconcile.Result{}, nil
+})
+
 // NewWatcher returns an implementation of Tracker that lets a Reconciler register a
 // particular resource as watching a resource for a particular lease duration.
 // This watch must be refreshed periodically (e.g. by a controller resync) or
 // it will expire.
-func NewWatcher(lease time.Duration, log logr.Logger, scheme *runtime.Scheme, enqueueTracked func(by client.Object, t Tracker) handler.EventHandler) Tracker {
-	tracker := New(lease, log)
-	return NewWatchingTracker(tracker, func(gvk schema.GroupVersionKind, controller controller.Controller) error {
+func NewWatcher(sm manager.SuperManager, lease time.Duration, log logr.Logger, scheme *runtime.Scheme, enqueueTracked func(by client.Object, t Tracker) handler.EventHandler) Tracker {
+	tracker := New(lease, log).(GKAwareTracker)
+	return NewWatchingTracker(tracker, func(gvk schema.GroupVersionKind) (context.CancelFunc, error) {
 		rObj, err := scheme.New(gvk)
 		obj := rObj.(crclient.Object)
 		if err != nil {
-			return err
+			return nil, err
 		}
 
-		return controller.Watch(&source.Kind{Type: obj}, enqueueTracked(obj, tracker))
+		// Create a new manager with its own cache which can be stopped when watches need to be stopped.
+		mgr, err := sm.NewManager()
```

Comment:
Can we share these managers/informer caches between trackers? Since a new tracker is created for every reconciler, if different resources watch the same duck types we'll end up with duplicate informer caches. It will make the eviction logic more complex, but as implemented I suspect this approach will consume more memory with duplicate informer caches than will be gained by garbage collecting stale informer caches.

Reply:
I will investigate.

Reply:
I'll need to pick your brains about the current design.

Reply:
Informers are now shared between watchers/trackers. See the design doc for details.